### DISCUSSION TITLE
Running a Single Application Through a VPN

### DISCUSSION THREAD

**Discussion by QasimK** [QasimK/qasimk.github.io#30] [Blog Posts]
Comment on [Running a Single Application Through a VPN](http://qasimk.io/2018/single-application-vpn/).

**Comment by Yureien** [comment]
Hey, I used this to run an app through a VPN, but that app requires access to an SQL server on the host running on localhost with port 5432. How do I access this?

**Comment by QasimK** [comment]
Hi @FadedCoder. That's a great question, and one that I hint at at the bottom of the article ("Aside: Local listening Apps). However, there are multiple solutions to this. The most reliable might involve setting up a private network for the network namespace like `10.0.3.0/24`, and then accessing the host via `10.0.3.1`. This is closer to what LXC or Docker can do with their containers.

A more hacky solution mentioned in the article is `socat`, which can connect arbitrary sockets (UDP, TCP, UNIX sockets) together. This is what I do because it ensures there is absolutely no other connection between the network namespace and the host.

While you may need to play around with it, doing something like

```
sudo /usr/bin/socat tcp-connect:127.0.0.1:5432,nofork exec:'ip netns exec NETNS socat STDIO tcp-listen\:127.0.0.1\:5432',fork,reuseaddr
```

This will create a listening socket on port 5432 inside the network namespace, which connects to port 5432 on the host.

I can't help you beyond this!

**Comment by Yureien** [comment]
I get a

```
2019/05/24 22:12:15 socat[1406] E parseopts(): option "nofork" not supported with this address type
```
when doing the `socat` one. Any help for that?

**Comment by QasimK** [comment]
Hey, I would suggest playing around with the command (maybe get rid of `nofork` from it?), and searching online for tutorials or examples of `socat`!
